lightbeam/validate.py

```diff
@@ -18,19 +33,190 @@ def __init__(self, lightbeam=None):
     # Validates (selected) endpoints
     def validate(self):
         self.lightbeam.api.load_swagger_docs()
-        asyncio.run(self.lightbeam.api.load_descriptors_values())
+        validation_methods = self.lightbeam.config.get("validate",{}).get("methods",self.DEFAULT_VALIDATION_METHODS)
+        if type(validation_methods)==str and (validation_methods=="*" or validation_methods.lower()=='all'):
+            validation_methods = self.DEFAULT_VALIDATION_METHODS
+            validation_methods.append("references")
+        self.logger.info(f"validating by methods {validation_methods}...")
+        if "descriptors" in validation_methods:
+            # load remote descriptors
+            asyncio.run(self.lightbeam.api.load_descriptors_values())
+            self.lightbeam.reset_counters()
+            self.load_local_descriptors()
+        
+        endpoints_with_data = self.lightbeam.get_endpoints_with_data()
+        self.lightbeam.endpoints = self.lightbeam.api.apply_filters(endpoints_with_data)
 
-        self.lightbeam.reset_counters()
+        # structures for local and remote reference lookups to prevent repeated lookups for the same thing
+        self.remote_reference_cache = {}
+        self.local_reference_cache = {}
 
-        local_descriptors = self.load_local_descriptors()
         for endpoint in self.lightbeam.endpoints:
-            if "Descriptor" in endpoint: swagger = self.lightbeam.api.descriptors_swagger
-            else: swagger = self.lightbeam.api.resources_swagger
-            self.validate_endpoint(swagger, endpoint, local_descriptors)
+            if "references" in validation_methods and "Descriptor" not in endpoint: # Descriptors have no references:
+                # We don't want every `do_validate_payload()` to separately have to open and scan
+                # local files looking for a matching payload; this pre-loads local data that
+                # might resolve references from within payloads of this endpoint.
+                # We assume that the data fits in memory; the largest Ed-Fi endpoints
+                # (studentSectionAssociations, studentSchoolAttendanceEvents, etc.) contain references
+                # to comparatively datasets (sections, schools, students).
+                self.build_local_reference_cache(endpoint)
+            asyncio.run(self.validate_endpoint(endpoint))
+    
+    def build_local_reference_cache(self, endpoint):
+        swagger = self.lightbeam.api.resources_swagger
+        definition = self.get_swagger_definition_for_endpoint(endpoint)
+        references_structure = self.load_references_structure(swagger, definition)
+        references_structure = self.rebalance_local_references_structure(references_structure)
```

Comment:
`load_references_structure()` is recursive, it calls itself (to handle potentially nested references). I'm not sure if we can combine these because of that?

Reply:
Yeah that's a fair point. I think there are ways around it (e.g. a `depth` parameter so the function knows whether it's running recursively or at top-level) but that may add more complexity than it's worth

Plus, I see on closer inspection that there is an instance where `load_references_structure` is called and `rebalance_local_references_structure` is not! So it seems this is a moot point!